// jokemaster/joke.mjs

const JOKE_DELIMITER = "\n";


class Joke{
    
    constructor(sourceJokes){
        this.jokeIndex;
        this.jokes = [];
        if(sourceJokes){
            if(Array.isArray(sourceJokes)){
                this.jokes = sourceJokes;    
            } else if(typeof sourceJokes === "string" ){
                this.jokes = Joke.#extractJokesFromString(sourceJokes);
            }
        }
    }

    getJokeIndex(){
       
        return this.jokeIndex = Math.floor(Math.random() * this.jokes.length-1) + 1 ;
    }

    tellAJoke(param){
        param = parseInt(param);
        let joke = "Sorry, I am all out of laughs";
        let index;

        if (param === -1) {
            index = this.getJokeIndex();
        }else{
            index = param;
        }
        if(this.jokes.length > 0){
            
            joke = this.jokes[index];
        }
        return [joke, index];
    }

    static #extractJokesFromString(source){
        const jokeArray = source.split(JOKE_DELIMITER);
        if(!Array.isArray(jokeArray) || jokeArray.length === 0){
            jokeArray = [];
        }
        return jokeArray;
    }

}


export { Joke as default, JOKE_DELIMITER };
